### PR TITLE
adding tensor_permutation example folder

### PR DESCRIPTION
**Pull request by arai713**

### CODE REVIEW

**Comment by poyenc** [comment]
Do you accidentally add the _.tensor_permute.cpp.swo_ file?

File: example/35_tensor_permute/tensor_permute.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    using btype = ck::remove_reference_t<decltype(B(0, 0, 0, 0))>;
+    for(std::size_t n = 0; n < shape[0]; ++n)
+        for(std::size_t c = 0; c < shape[1]; ++c)
+            for(std::size_t h = 0; h < shape[2]; ++h)
+                for(std::size_t w = 0; w < shape[3]; ++w)
+                {
+                    auto a_val  = A(n, c, h, w);
+                    btype b_val = 0;
+                    functor(b_val, a_val);
+                    B(n, h, w, c) = b_val;
```
Comment (zjing14) [inline example/35_tensor_permute/tensor_permute.cpp:44]: functor(B(n, h, w, c), a_val);

**Comment by zjing14** [comment]
Please rename the example folder to "36_elementwise_permute".  "35" has been used already. 
Please add numDim and data type into the example file name (e.g., example/36_elementwise_permute/elementwise_permute_4D_fp16.cpp)

File: example/35_tensor_permute/tensor_permute.cpp [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+template <typename HostTensorA, typename HostTensorB, typename Functor>
+void host_elementwise4D(HostTensorB& B,
+                        const HostTensorA& A,
+                        const std::vector<std::size_t>& shape,
+                        Functor functor)
+{
+    using btype = ck::remove_reference_t<decltype(B(0, 0, 0, 0))>;
+    for(std::size_t n = 0; n < shape[0]; ++n)
+        for(std::size_t c = 0; c < shape[1]; ++c)
```
Comment (zjing14) [inline example/35_tensor_permute/tensor_permute.cpp:37]: Inconsistent here. Input "shape" is nhwc, but c = shape[1] here.

Comment (zjing14) [reply example/35_tensor_permute/tensor_permute.cpp:37]: Change "shape" to "shape_nhwc" for better readability.

File: example/35_tensor_permute/tensor_permute.cpp [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    ck::tensor_operation::device::DeviceElementwise<ck::Tuple<ADataType>,
+                                                    ck::Tuple<BDataType>,
+                                                    PassThrough,
+                                                    4,
+                                                    8,
+                                                    ck::Sequence<8>,
+                                                    ck::Sequence<1>>;
+
+template <typename HostTensorA, typename HostTensorB, typename Functor>
+void host_elementwise4D(HostTensorB& B,
```
Comment (zjing14) [inline example/35_tensor_permute/tensor_permute.cpp:30]: Use "B_nhwc" and "A_nchw" for better readabililty.

**Comment by zjing14** [comment]
@arai713 Please fix comments above.

**Comment by poyenc** [comment]
Forgot to add **_36_elementwise_permute_** subdirectory into _example/CmakeLists.txt_?

File: example/CMakeLists.txt [hunk 4]
```diff
@@ -21,10 +21,12 @@ function(add_example_executable_no_testing EXAMPLE_NAME FILE_NAME)
     add_dependencies(examples ${EXAMPLE_NAME})
 endfunction(add_example_executable_no_testing EXAMPLE_NAME)
 
+
```
Comment (zjing14) [inline example/CMakeLists.txt:24]: Please remove the change.

File: example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    std::array<ck::index_t, 4> ab_lengths;
+    std::array<ck::index_t, 4> a_strides = {static_cast<int>(nchw[1] * nchw[2] * nchw[3]),
+                                            static_cast<int>(nchw[2] * nchw[3]),
+                                            static_cast<int>(nchw[3]),
+                                            1};
+    std::array<ck::index_t, 4> b_strides = {
+        static_cast<int>(nhwc[1] * nhwc[2] * nhwc[3]), 1, 32, 4};
+
+    // std::cout << "Length: " << ab_lengths << std::endl;
```
Comment (zjing14) [inline example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp:74]: Please remove unused lines.

File: example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    // std::cout << "A stride: " << a_strides << std::endl;
+    // std::cout << "B stride: " << b_strides << std::endl;
+
+    std::copy(nchw.begin(), nchw.end(), ab_lengths.begin());
+    // std::copy(a.mDesc.GetStrides().begin(), a.mDesc.GetStrides().end(), a_strides.begin());
```
Comment (zjing14) [inline example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp:79]: Same here.

File: example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    std::cout << "Perf: " << ave_time << " ms" << std::endl;
+
+    bool pass = true;
+
+    if(do_verification)
+    {
+        b_device_buf.FromDevice(b.mData.data());
+        LogRangeAsType<float>(std::cout << "Tensor b  : ", b.mData, ",") << std::endl;
+        Tensor<BDataType> host_b(nhwc);
+        host_elementwise4D<Tensor<ADataType>, Tensor<BDataType>, PassThrough>(
```
Comment (zjing14) [inline example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp:104]: `<Tensor<ADataType>, Tensor<BDataType>, PassThrough>` is unnecessary.

File: example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                                                    PassThrough,
+                                                    4,
+                                                    8,
+                                                    ck::Sequence<1>,
+                                                    ck::Sequence<1>>;
+
+template <typename HostTensorA, typename HostTensorB, typename Functor>
+void host_elementwise4D(HostTensorB& B_nhwc,
+                        const HostTensorA& A_nchw,
+                        const std::vector<std::size_t>& shape_nchw,
```
Comment (zjing14) [inline example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp:32]: `shape_nchw` is unncessary.

You can get dims by "A_nchw.mDesc.GetLength[Ix];"

For example, 

```
const auto N = A_nchw.mDesc.GetLengths()[0];
const auto C = A_nchw.mDesc.GetLengths()[1];
const auto H = A_nchw.mDesc.GetLengths()[2];
const auto W = A_nchw.mDesc.GetLengths()[3];
```

File: example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+using BDataType = F16;
+
+using PassThrough = ck::tensor_operation::element_wise::PassThrough;
+using DeviceElementwisePermuteInstance =
+    ck::tensor_operation::device::DeviceElementwise<ck::Tuple<ADataType>,
+                                                    ck::Tuple<BDataType>,
+                                                    PassThrough,
+                                                    4,
+                                                    8,
+                                                    ck::Sequence<1>,
```
Comment (zjing14) [inline example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp:26]: use vector load for source (i.e., ck::Sequence<8>;).

**Review by zjing14 — APPROVED** [review]
LGTM

File: example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+    std::array<const void*, 1> input = {a_device_buf.GetDeviceBuffer()};
+    std::array<void*, 1> output      = {b_device_buf.GetDeviceBuffer()};
+
+    std::array<ck::index_t, 4> ab_lengths;
+    std::array<ck::index_t, 4> a_strides = {static_cast<int>(nchw[1] * nchw[2] * nchw[3]),
+                                            static_cast<int>(nchw[2] * nchw[3]),
+                                            static_cast<int>(nchw[3]),
+                                            1};
+    std::array<ck::index_t, 4> b_strides = {static_cast<int>(nhwc[1] * nhwc[2] * nhwc[3]),
```
Comment (zjing14) [inline example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp:68]: Should use the same length (i.e., nchw[]) to calculate strides.

Comment (zjing14) [reply example/38_elementwise_permute/elementwise_permute_4D_fp16.cpp:68]: ab_lengths = {N, H, W, C}, a_strides = {C * H * W, W, 1, H * W}, b_stride = {H * W * C, W * C, C, 1}

**Pull request merged**